test/model/autocomplete_test.dart

```diff
@@ -450,22 +540,120 @@ void main() {
+        test('StreamNarrow', () async {
+          await checkInitialResultsIn(streamNarrow, expected: [4, 0, 3, 1, 2]);
+        });
+
+        test('StreamNarrow, new message arrives', () async {
+          await checkInitialResultsIn(streamNarrow, expected: [4, 0, 3, 1, 2]);
+
+          // Until now, latest message id in [stream] is 60.
+          await store.addMessage(streamMessage(id: 70, senderId: 2));
+
+          await checkResultsIn(streamNarrow, expected: [2, 4, 0, 3, 1]);
+        });
+
+        test('StreamNarrow, a batch of older messages arrives', () async {
+          await checkInitialResultsIn(streamNarrow, expected: [4, 0, 3, 1, 2]);
+
+          // Until now, oldest message id in [stream] is 50.
+          final oldMessages = [
+            streamMessage(id: 30, senderId: 1),
+            streamMessage(id: 40, senderId: 2),
+          ];
+          connection.prepare(json: olderResult(
+            anchor: 50, foundOldest: false,
+            messages: oldMessages,
+          ).toJson());
+          await messageList.fetchOlder();
+
+          await checkResultsIn(streamNarrow, expected: [4, 0, 2, 1, 3]);
+        });
+
+        test('TopicNarrow, no other messages are in stream', () async {
+          await checkInitialResultsIn(topicNarrow, expected: [0, 3, 1, 2, 4]);
+        });
+
+        test('TopicNarrow, other messages are in stream', () async {
+          await checkInitialResultsIn(topicNarrow, expected: [0, 4, 3, 1, 2],
+            includeStream: true);
+        });
+
+        test('TopicNarrow, new message arrives', () async {
+          await checkInitialResultsIn(topicNarrow, expected: [0, 3, 1, 2, 4]);
+
+          // Until now, latest message id in [topic] is 50.
+          await store.addMessage(streamMessage(id: 60, senderId: 2, topic: topic));
+
+          await checkResultsIn(topicNarrow, expected: [2, 0, 3, 1, 4]);
+        });
+
+        test('TopicNarrow, a batch of older messages arrives', () async {
+          await checkInitialResultsIn(topicNarrow, expected: [0, 3, 1, 2, 4]);
+
+          // Until now, oldest message id in [topic] is 50.
+          final oldMessages = [
+            streamMessage(id: 30, senderId: 2, topic: topic),
+            streamMessage(id: 40, senderId: 4, topic: topic),
+          ];
+          connection.prepare(json: olderResult(
+            anchor: 50, foundOldest: false,
+            messages: oldMessages,
+          ).toJson());
+
+          await messageList.fetchOlder();
+          await checkResultsIn(topicNarrow, expected: [0, 4, 2, 3, 1]);
+        });
       });
 
-      test('DmNarrow', () async {
-        await checkResultsIn(
-          DmNarrow.withUser(eg.selfUser.userId, selfUserId: eg.selfUser.userId),
-          expected: [3, 0, 1, 2, 4],
-        );
+      group('DmNarrow', () {
+        test('DmNarrow, with no topic/stream message history', () async {
+          await prepareStore();
+          await checkResultsIn(dmNarrow, expected: [3, 0, 1, 2, 4]);
```

Comment:
It's hard to interpret most of the test cases in this "suggests relevant users" group — for example, why is 3, 0, 1, 2, 4 the correct result to expect here, and why is 0, 3, 1, 2, 4 the correct result to expect in some of the tests above? The details that cause the right answers to be those lists and not something else are separated from here by about 100 lines of code, and they require some careful working-through to verify.

In general it's best for a test to be self-contained — all the key details that tell the story of the test, and determine why one value is expected instead of another, live in the test case's own source code, or failing that very close above it. That way
* a reader can look at the test and confirm it's checking for the right answer instead of a wrong answer;
* a reader can look at the test and understand what it's testing, and look at the ensemble of tests to think about things that aren't yet tested;
* in the future when changing something about the code under test and the test breaks, a reader can understand what the test was trying to say so they can decide whether the failure signals a bug in their change or is an intended effect of their change, and so they can update the test if necessary so it continues to check what it was intended to check.

We use a lot of test helper functions to enable test cases to be concise and focused, because that also aids in all three of the points above. But when doing so, the key is to move out the boring details (the things that can be summarized as the helpers just doing things that ought to be done), and keep in the details that are part of the test's story, or that are necessary in order to make sense of other details in the test. So for example as long as this ranking `[3, 0, 1, 2, 4]` is here, the details of `messages` and `dmConversations` that give rise to it should be here or near here too.